private/bufpkg/bufcheck/buflint/internal/buflintcheck/buflintcheck.go

```diff
@@ -963,3 +968,95 @@ func checkSyntaxSpecified(add addFunc, file protosource.File) error {
 	}
 	return nil
 }
+
+var (
+	// fieldConstraintsDesc provides a Descriptor for validate.FieldConstraints.
+	fieldConstraintsDesc = (*validate.FieldConstraints)(nil).ProtoReflect().Descriptor()
+
+	// fieldConstraintsOneOfDesc provides the OneOfDescriptor for the type union
+	// in FieldConstraints.
+	fieldConstraintsOneOfDesc = fieldConstraintsDesc.Oneofs().ByName("type")
+)
+
+// CheckValidateRulesTypesMatch is a check function.
+var CheckValidateRulesTypesMatch = newMessageCheckFunc(checkValidateRulesTypesMatch)
+
+func checkValidateRulesTypesMatch(add addFunc, message protosource.Message) error {
+	for _, field := range message.Fields() {
+		data, err := getDataByExtension(field, validate.E_Field)
+		if err != nil {
+			return err
+		}
+		var constraints validate.FieldConstraints
+		if err := proto.Unmarshal(data, &constraints); err != nil {
+			return fmt.Errorf("unmarshal error for field %q: %v", field.FullName(), err)
+		}
+		constraintsRefl := constraints.ProtoReflect()
+		setOneOf := constraintsRefl.WhichOneof(fieldConstraintsOneOfDesc)
+		if setOneOf == nil {
+			continue
+		}
+
+		got := field.Type()
+		expected := getTypeName(setOneOf.Name())
+		if got != expected {
+			add(field, field.Location(), nil, "constraint type mismatch on %q, expected %q but got %s", field.FullName(), expected, got)
+		}
+	}
+	return nil
+}
+
+func getDataByExtension(field protosource.OptionExtensionDescriptor, extensionType protoreflect.ExtensionType) ([]byte, error) {
+	extension, ok := field.OptionExtension(extensionType)
+	if !ok {
+		return nil, nil
+	}
+	fieldConstraints, ok := extension.(*dynamicpb.Message)
+	if !ok {
+		return nil, fmt.Errorf("unexpected extension type for field got %T", extension)
+	}
+	data, err := proto.Marshal(fieldConstraints)
```

Comment:
Why do we need to marshal (and the unmarshal) here? I would think we could return `*dynamicpb.Message` as `proto.Message` directly, and then use it for i.e. calling `ProtoReflect()`, is that not possible?

Reply:
Can we add some comments as to what this function is doing, so future readers can quickly get context? I know the private functions of `bufcheck` aren't the best-documented, but this function is particularly confusing at least to me as an external reader. Some comments about why we are doing the marshal/unmarshal as well would be helpful.